### PR TITLE
Multitenancy alternate implementation

### PR DESCRIPTION
**Pull request by nrempel**
This is a proof of concept for a slightly different approach for managing multiple wallets.

### CODE REVIEW

File: aries_cloudagent/admin/server.py [hunk 1]
```diff
@@ -263,6 +265,27 @@ async def collect_stats(request, handler):
...
+        @web.middleware
+        async def set_wallet(request, handler):
+            wallet_id = request.headers.get("x-wallet-id")
+            if wallet_id:
+                wallet_record = await WalletRecord.retrieve_by_id(
+                    self.context, wallet_id
+                )
+                # We amend the context settings so that the wallet that will be
+                # opened later has different context
+                self.context.settings = self.context.settings.extend(
```
Comment (nrempel) [inline aries_cloudagent/admin/server.py:277]: Modify the settings for the request context. This way any wallets that are injected after this point will use the modified wallet configuration and open different wallets.

File: aries_cloudagent/config/default_context.py [hunk 2]
```diff
@@ -75,7 +75,8 @@ async def bind_providers(self, context: InjectionContext):
                         # "unpack_message",
                         "get_local_did",
                     ),
-                )
+                ),
+                ("wallet.name",),
```
Comment (nrempel) [inline aries_cloudagent/config/default_context.py:79]: Specify what subset of the settings should be used to determine uniqueness

File: aries_cloudagent/config/provider.py [hunk 3]
```diff
@@ -70,18 +72,29 @@ async def provide(self, config: BaseSettings, injector: BaseInjector):
 class CachedProvider(BaseProvider):
```
Comment (nrempel) [inline aries_cloudagent/config/provider.py:72]: Keep single CachedProvider but make it a little smarter

File: aries_cloudagent/config/provider.py [hunk 4]
```diff
@@ -70,18 +72,29 @@ async def provide(self, config: BaseSettings, injector: BaseInjector):
...
-        if not self._instance:
-            self._instance = await self._provider.provide(config, injector)
-        return self._instance
+        """
+        Provide the object instance given a config and injector.
+        
+        Instances are cached keyed on a SHA256 digest of the relevant subset
+        of settings.
+        """
+        instance_vals = {key: config.get(key) for key in self._unique_settings_keys}
```
Comment (nrempel) [inline aries_cloudagent/config/provider.py:90]: Key the cached provider off a hash of the provided subset of settings

File: aries_cloudagent/wallet/routes.py [hunk 5]
```diff
@@ -86,6 +121,82 @@ def format_did_info(info: DIDInfo):
...
+    else:
+        raise web.HTTPBadRequest(reason="Name needs to be provided to create a wallet.")
+    config["key"] = body.get("wallet_key")
+    wallet_type = body.get("wallet_type")
+    if wallet_type not in WALLET_TYPES:
+        raise web.HTTPBadRequest(reason="Specified wallet type is not supported.")
+    config["type"] = wallet_type
+
+    wallet_record = WalletRecord(wallet_config=config)
+    await wallet_record.save(context)
```
Comment (nrempel) [inline aries_cloudagent/wallet/routes.py:153]: Wallet configurations are stored as wallet records

File: aries_cloudagent/wallet/routes.py [hunk 6]
```diff
@@ -86,6 +121,82 @@ def format_did_info(info: DIDInfo):
...
+    context = request.app["request_context"]
+
+    try:
+        records = await WalletRecord.query(context)
+        results = [record.serialize() for record in records]
+    except StorageNotFoundError:
+        raise web.HTTPNotFound()
+    return web.json_response(results)
+
+
```
Comment (nrempel) [inline aries_cloudagent/wallet/routes.py:199]: Todo:

wallet delete, rekey, etc

File: aries_cloudagent/wallet/routes.py [hunk 7]
```diff
@@ -16,11 +16,28 @@
...
 from ..messaging.valid import ENDPOINT, INDY_CRED_DEF_ID, INDY_DID, INDY_RAW_PUBLIC_KEY
+from ..wallet.models.wallet_record import WalletRecord
+from ..storage.error import StorageNotFoundError
+from .models.wallet_record import WalletRecordSchema
 
 from .base import DIDInfo, BaseWallet
 from .error import WalletError, WalletNotFoundError
 
 
+WALLET_TYPES = {
```
Comment (sklump) [inline aries_cloudagent/wallet/routes.py:27]: What would you think of an Enum here?

File: aries_cloudagent/wallet/models/wallet_record.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        return super().__eq__(other)
+
+
+class WalletRecordSchema(BaseExchangeSchema):
+    """Schema to allow serialization/deserialization of record."""
+
+    class Meta:
+        """WalletRecordSchema metadata."""
+
+        model_class = WalletRecord
```
Comment (sklump) [inline aries_cloudagent/wallet/models/wallet_record.py:68]:
```
from marshmallow import EXCLUDE, ...
...
    class Meta:
        model_class = WalletRecord
        unknown = EXCLUDE
```

File: aries_cloudagent/wallet/routes.py [hunk 9]
```diff
@@ -76,6 +93,24 @@ class CredDefIdMatchInfoSchema(Schema):
     )
 
 
+class CreateWalletRequestSchema(Schema):
+    """Request schema for adding a new wallet which will be registered by the agent."""
+
+    wallet_name = fields.Str(description="Wallet identifier.", example="MyNewWallet")
```
Comment (sklump) [inline aries_cloudagent/wallet/routes.py:99]: Should be a pick-list, generate from allowable types; e.g.,
```
from marshmallow import ..., validate
...
    wallet_name = fields.Str(..., validate=validate.OneOf([wallet_type for wallet_type in WALLET_TYPES]))
``

File: aries_cloudagent/wallet/routes.py [hunk 10]
```diff
@@ -86,6 +121,82 @@ def format_did_info(info: DIDInfo):
...
+    body = await request.json()
+
+    config = {}
+    if body.get("wallet_name"):
+        config["name"] = body.get("wallet_name")
+    else:
+        raise web.HTTPBadRequest(reason="Name needs to be provided to create a wallet.")
+    config["key"] = body.get("wallet_key")
+    wallet_type = body.get("wallet_type")
+    if wallet_type not in WALLET_TYPES:
```
Comment (sklump) [inline aries_cloudagent/wallet/routes.py:148]: By limiting choices above, this check becomes redundant

**Pull request closed without merging**